src/execution_plan/ops/op_expand_into.c

```diff
@@ -7,67 +7,124 @@
+        offset += Edge_ToString(op->exp->edge, buff + offset, buff_len - offset);
+        offset += snprintf(buff + offset, buff_len-offset, "->");
+    } else {
+        offset += snprintf(buff + offset, buff_len-offset, "->");
+    }
+    offset += Node_ToString(op->exp->dest_node, buff + offset, buff_len - offset);
+    return offset;
+}
+
 /* Expand into checks to see if two resolved nodes are connected
  * MATCH (a)-[]->(b), (a)-[:e]->(b) 
  * in the above we might figure out that indeed a is connected somehow to b
  * (a)-[]->(b) 
  * we still need to make sure a is connected via an e edge to b
- * as a and b are already resolved to just need to see if they are 
+ * as a and b are already resolved we just need to see if they are 
  * connected via an e edge. */
-
-OpBase* NewExpandIntoOp(Node *a, Node *b, Edge *e, AST *ast) {
+OpBase* NewExpandIntoOp(AlgebraicExpression *exp, uint srcRecIdx, uint destRecIdx, uint edgeRecIdx) {
     OpExpandInto *expandInto = malloc(sizeof(OpExpandInto));
-    expandInto->srcRecIdx = AST_GetAliasID(ast, a->alias);
-    expandInto->destRecIdx = AST_GetAliasID(ast, b->alias);
-    expandInto->gc = GraphContext_GetFromTLS();
-    expandInto->e = e;
-    expandInto->edges = array_new(Edge*, 0);
+    expandInto->r = NULL;
+    expandInto->exp = exp;
+    expandInto->srcRecIdx = srcRecIdx;
+    expandInto->destRecIdx = destRecIdx;
+    expandInto->edgeRecIdx = edgeRecIdx;
+    expandInto->edges = NULL;
+    expandInto->g = GraphContext_GetFromTLS()->g;
+    expandInto->relation = (exp->edge) ? Edge_GetRelationID(exp->edge) : GRAPH_NO_RELATION;
 
     // Set our Op operations
     OpBase_Init(&expandInto->op);
     expandInto->op.name = "Expand Into";
     expandInto->op.type = OPType_EXPAND_INTO;
     expandInto->op.consume = OpExpandIntoConsume;
     expandInto->op.reset = OpExpandIntoReset;
+    expandInto->op.toString = ExpandIntoToString;
     expandInto->op.free = OpExpandIntoFree;
-    expandInto->op.modifies = NewVector(char*, 0);
+
+    if(exp->edge) {
+        expandInto->op.modifies = NewVector(char*, 1);
+        Vector_Push(expandInto->op.modifies, exp->edge->alias);
+        expandInto->edges = array_new(Edge, 32);
+    }
+
     return (OpBase*)expandInto;
 }
 
 Record OpExpandIntoConsume(OpBase *opBase) {
+    Node *a;
+    Node *b;
+    NodeID aID;
+    NodeID bID;
     OpExpandInto *op = (OpExpandInto*)opBase;
     OpBase *child = op->op.children[0];
-    Record r = NULL;
 
-    // Clear edges array, see if a is connected to b via e.
-    array_clear(op->edges);
-    while(array_len(op->edges) == 0) {
+    /* If we're required to update edge,
+     * try to get an edge, if successful we can return quickly,
+     * otherwise try to get a new pair of source and destination nodes. */
+    if(op->exp->edge) {
+        if(_SetEdge(op, op->r)) {
+            return Record_Clone(op->r);
+        } else {
+            // No more edges to consume, free previous record.
```

Comment:
Should we be freeing here and in the `if` block?

Reply:
I think we should free op->r if we didn't managed to get an edge,
same thing should happen in op_conditional_traverse.
Need to double check.